src/mon/OSDMonitor.cc

```diff
@@ -2668,6 +2673,56 @@ bool OSDMonitor::prepare_pg_created(MonOpRequestRef op)
   return true;
 }
 
+bool OSDMonitor::preprocess_pg_ready_to_merge(MonOpRequestRef op)
+{
+  op->mark_osdmon_event(__func__);
+  auto m = static_cast<MOSDPGReadyToMerge*>(op->get_req());
+  dout(10) << __func__ << " " << *m << dendl;
+  auto session = m->get_session();
+  if (!session) {
+    dout(10) << __func__ << ": no monitor session!" << dendl;
+    return true;
+  }
+  if (!session->is_capable("osd", MON_CAP_X)) {
+    derr << __func__ << " received from entity "
+         << "with insufficient privileges " << session->caps << dendl;
+    return true;
+  }
+  const pg_pool_t *pi = osdmap.get_pg_pool(m->pgid.pool());
+  if (!pi) {
+    derr << __func__ << " pool for " << m->pgid << " dne" << dendl;
+    return true;
+  }
+  if (pi->get_pg_num() <= m->pgid.ps()) {
+    dout(20) << " pg_num " << pi->get_pg_num() << " already < " << m->pgid << dendl;
+    return true;
+  }
+  if (pi->get_pg_num_pending() > m->pgid.ps()) {
+    dout(20) << " pg_num_pending " << pi->get_pg_num_pending() << " > " << m->pgid << dendl;
+    return true;
+  }
```

Comment:
I think we don't want that. When we get ReadyToMerge it means we have quiesced updates and stabilized all clean replicas at the same update, which means we'll be able to do a clean merge.  Doesn't matter if we've raced with a pg location change or whatever.

We do need an epoch check, though, to make sure the message doesn't correspond to a previous interval of pg_num_pending == pg_num-1; i'll add that!

My plan is to make the merge in the case where the last_update doesn't line up just set last_backfill = hobject_t() too.  This might happen if, say, you just went active+clean after a backfill and an old stray copy hasn't been removed yet.

Still trying to figure out what safety "outs" we can build in in case the merge goes wrong (probably by intervening with ceph-objectstore-tool?)

Reply:
We can check that it was the primary OSD at the epoch in question too. I realize we pretty much let the OSDs do whatever they want but that seems like a good sanity check for something like this (if there are bugs between the OSDs and monitors, it will be super bad!) and also a way to start sneaking some level of security in. :p